app/views/admin/api_keys.py

```diff
@@ -0,0 +1,225 @@
+                'full_read_permission': utils.validate_checkbox_as_bool,
+                'is_valid': utils.validate_checkbox_as_bool,
+                'key': utils.strip,
+                'mark_notes_reviewed': utils.validate_checkbox_as_bool,
+                'organization_name': utils.strip,
+                'read_permission': utils.validate_checkbox_as_bool,
+                'search_permission': utils.validate_checkbox_as_bool,
+                'stats_permission': utils.validate_checkbox_as_bool,
+                'subscribe_permission': utils.validate_checkbox_as_bool,
+            })
+
+    def get(self, request, *args, **kwargs):
+        if self.params.get('log_key'):
+            management_log_key = self.params.log_key
+            management_log = db.get(management_log_key)
+            message = ''
+            if management_log.action == model.ApiKeyManagementLog.CREATE:
+                message = t.ugettext(
+                    'A new API key has been created successfully.')
+            elif management_log.action == model.ApiKeyManagementLog.UPDATE:
+                message = t.ugettext(
+                    'The API key has been updated successfully.')
+            return self._render_form(management_log.authorization, message)
+        else:
+            return self._render_form()
+
+    def post(self, request, *args, **kwargs):
+        if not (self.params.get('xsrf_token') and self.xsrf_tool.verify_token(
+            self.params.xsrf_token, self.env.user.user_id(), 'admin_api_keys')):
+            return self.error(403)
+
+        # Navigation to an individual key's management page is handled by making
+        # a POST request to this view. When it's such a request the edit_form
+        # param will be set.
+        if self.params.get('edit_form'):
+            authorization = db.get(self.params.get('authorization_key'))
+            if not authorization:
+                return self.error(404, t.ugettext(
+                    'No such Authorization entity.'))
+            return self._render_form(authorization)
+
+        if not (self.params.get('contact_name') and
+                self.params.get('contact_email') and
+                self.params.get('organization_name')):
+            return self.error(400, t.ugettext(
+                'Please fill in all the required fields.'))
+
+        repo = self.env.repo or '*'
+        if self.params.get('key'):
+            # Just override the existing one.
+            existing_authorization = db.get(self.params.get('key'))
+            if not existing_authorization:
+                return self.error(404, t.ugettext(
+                    'No such Authorization entity.'))
+            # This shouldn't happen unless an admin does something funny with
+            # URLs, but check just to be safe.
+            if existing_authorization.repo != repo:
+                return self.error(400, t.ugettext(
+                    'Authorization already exists for another repo!'
```

Comment:
nit: missing a space between the two sentences?

Reply:
Fixed.